File: python/examples/drawing/render.py
import tkinter as tk

from schema import Style, Box, Ellipse, Arrow, Drawing, UnknownText


# Map line style to dash patterns
dash_pattern = {
    "solid": "",
    "dashed": (4, 4),  # 4 pixels drawn, 4 pixels space
    "dotted": (1, 1),  # 1 pixel drawn, 1 pixel space
}


def render_drawing(drawing: Drawing):
    window = tk.Tk()
    window.title("Drawing")
    window.configure(bg="white")

    canvas = tk.Canvas(window, width=800, height=600, bg="white", highlightthickness=0)
    canvas.pack(padx=10, pady=10)

    def draw_box(box: Box):
        x1, y1 = box.x, box.y
        x2, y2 = x1 + box.width, y1 + box.height
        canvas.create_rectangle(
            x1,
            y1,
            x2,
            y2,
            outline=getattr(box.style, "line_color", None) or "black",
            fill=getattr(box.style, "fill_color", None) or "",
        )
        if box.text:
            canvas.create_text((x1 + x2) / 2, (y1 + y2) / 2, text=box.text, fill="black")

    def draw_ellipse(ellipse: Ellipse):
        x1, y1 = ellipse.x, ellipse.y
        x2, y2 = x1 + ellipse.width, y1 + ellipse.height
        canvas.create_oval(
            x1,
            y1,
            x2,
            y2,
            outline=getattr(ellipse.style, "line_color", None) or "black",
            fill=getattr(ellipse.style, "fill_color", None) or "",
        )
        if ellipse.text:
            canvas.create_text((x1 + x2) / 2, (y1 + y2) / 2, text=ellipse.text, fill="black")

    def draw_arrow(arrow: Arrow):
        canvas.create_line(
            arrow.start_x,
            arrow.start_y,
            arrow.end_x,
            arrow.end_y,
            dash=dash_pattern[getattr(arrow.style, "line_style", None) or "solid"],
            arrow=tk.LAST,
            fill=getattr(arrow.style, "line_color", None) or "black",
        )

    for item in drawing.items:
        match item:
            case Box():
                draw_box(item)
            case Ellipse():
                draw_ellipse(item)
            case Arrow():
                draw_arrow(item)
            case UnknownText():
                print(f"Unknown text: {item.text}")

    window.mainloop()


if __name__ == "__main__":
    example_drawing = Drawing(
        type="Drawing",
        items=[
            Box(
                type="Box",
                x=50,
                y=50,
                width=100,
                height=100,
                text="Hello",
                style=Style(type="Style"),
            ),
            Ellipse(
                type="Ellipse",
                x=200,
                y=50,
                width=150,
                height=100,
                text="World",
                style=Style(type="Style", fill_color="lightblue"),
            ),
            Arrow(
                type="Arrow",
                start_x=50,
                start_y=200,
                end_x=150,
                end_y=200,
                style=Style(type="Style", line_style="dashed"),
            ),
        ],
    )

    render_drawing(example_drawing)

File: typescript/examples/drawing/src/render.ts
import { Arrow, Box, Drawing, Ellipse, Style } from "./drawingSchema";

function escapeXml(text: string): string {
    return text
        .replaceAll("&", "&amp;")
        .replaceAll("<", "&lt;")
        .replaceAll(">", "&gt;")
        .replaceAll("\"", "&quot;")
        .replaceAll("'", "&apos;");
}

function lineDash(style?: Style): string | undefined {
    switch (style?.line_style) {
        case "dashed":
            return "6 4";
        case "dotted":
            return "2 3";
        default:
            return undefined;
    }
}

function stroke(style?: Style): string {
    return style?.line_color ?? "black";
}

function strokeWidth(style?: Style): number {
    return style?.line_thickness ?? 1;
}

function fill(style?: Style): string {
    return style?.fill_color ?? "none";
}

function renderBox(item: Box): string {
    const rounded = item.style?.corners === "rounded";
    const text = item.text ? `<text x="${item.x + item.width / 2}" y="${item.y + item.height / 2}" text-anchor="middle" dominant-baseline="middle">${escapeXml(item.text)}</text>` : "";
    return `<rect x="${item.x}" y="${item.y}" width="${item.width}" height="${item.height}"${rounded ? " rx=\"8\" ry=\"8\"" : ""} stroke="${stroke(item.style)}" stroke-width="${strokeWidth(item.style)}" fill="${fill(item.style)}" />${text}`;
}

function renderEllipse(item: Ellipse): string {
    const cx = item.x + item.width / 2;
    const cy = item.y + item.height / 2;
    const text = item.text ? `<text x="${cx}" y="${cy}" text-anchor="middle" dominant-baseline="middle">${escapeXml(item.text)}</text>` : "";
    return `<ellipse cx="${cx}" cy="${cy}" rx="${item.width / 2}" ry="${item.height / 2}" stroke="${stroke(item.style)}" stroke-width="${strokeWidth(item.style)}" fill="${fill(item.style)}" />${text}`;
}

function renderArrow(item: Arrow): string {
    const dash = lineDash(item.style);
    return `<line x1="${item.start_x}" y1="${item.start_y}" x2="${item.end_x}" y2="${item.end_y}" stroke="${stroke(item.style)}" stroke-width="${strokeWidth(item.style)}"${dash ? ` stroke-dasharray="${dash}"` : ""} marker-end="url(#arrowhead)" />`;
}

function getCanvasSize(drawing: Drawing): { width: number; height: number } {
    let maxX = 800;
    let maxY = 600;
    for (const item of drawing.items) {
        switch (item.type) {
            case "Box":
            case "Ellipse":
                maxX = Math.max(maxX, item.x + item.width + 40);
                maxY = Math.max(maxY, item.y + item.height + 40);
                break;
            case "Arrow":
                maxX = Math.max(maxX, item.start_x + 40, item.end_x + 40);
                maxY = Math.max(maxY, item.start_y + 40, item.end_y + 40);
                break;
        }
    }
    return { width: maxX, height: maxY };
}

export function renderDrawingToSvg(drawing: Drawing): string {
    const { width, height } = getCanvasSize(drawing);
    const shapes = drawing.items.flatMap((item) => {
        switch (item.type) {
            case "Box":
                return [renderBox(item)];
            case "Ellipse":
                return [renderEllipse(item)];
            case "Arrow":
                return [renderArrow(item)];
            default:
                return [];
        }
    });

    return [
        `<svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}" viewBox="0 0 ${width} ${height}">`,
        "<defs>",
        "<marker id=\"arrowhead\" markerWidth=\"10\" markerHeight=\"7\" refX=\"9\" refY=\"3.5\" orient=\"auto\">",
        "<polygon points=\"0 0, 10 3.5, 0 7\" fill=\"black\" />",
        "</marker>",
        "</defs>",
        "<rect width=\"100%\" height=\"100%\" fill=\"white\" />",
        ...shapes,
        "</svg>",
    ].join("\n");
}
